lib/miq_expression.rb

```diff
@@ -1554,36 +1554,43 @@ def self._quick_search?(e)
     end
   end
 
-  def custom_attribute_columns_and_models(expression = nil)
-    return custom_attribute_columns_and_models(exp).uniq if expression.nil?
+  def custom_attribute_columns_and_models
+    @virtual_custom_attributes ||= begin
+      @virtual_custom_attributes = []
+      _custom_attribute_columns_and_models
+      @virtual_custom_attributes
+    end
+  end
+
+  def _custom_attribute_columns_and_models(expression = nil)
+    return _custom_attribute_columns_and_models(exp) if expression.nil?
```

Comment:
Perhaps this could be passed in by the caller instead of being the default case here? The former is more intention-revealing. This being a guard clause makes me stop for a second to consider the reasons this might be `nil` - I assume that is the only reason?